provider/gce/environ_network.go

```diff
@@ -1,40 +1,213 @@
-// Copyright 2014 Canonical Ltd.
+// Copyright 2017 Canonical Ltd.
 // Licensed under the AGPLv3, see LICENCE file for details.
 
 package gce
 
 import (
 	"github.com/juju/errors"
+	"github.com/juju/utils/set"
+	"gopkg.in/juju/names.v2"
 
+	"github.com/juju/juju/instance"
 	"github.com/juju/juju/network"
-	"github.com/juju/juju/provider/common"
 )
 
-// globalFirewallName returns the name to use for the global firewall.
-func (env *environ) globalFirewallName() string {
-	return common.EnvFullName(env.uuid)
+// Subnets implements environs.NetworkingEnviron.
+func (e *environ) Subnets(inst instance.Id, subnetIds []network.Id) ([]network.SubnetInfo, error) {
+	// In GCE all the subnets are in all AZs.
+	zones, err := e.AvailabilityZones()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	zoneNames := make([]string, len(zones))
+	for i, zone := range zones {
+		zoneNames[i] = zone.Name()
+	}
+
+	ids := makeIncludeSet(subnetIds)
+	var results []network.SubnetInfo
+	if inst == instance.UnknownId {
+		results, err = e.getMatchingSubnets(ids, zoneNames)
+	} else {
+		results, err = e.getInstanceSubnets(inst, ids, zoneNames)
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if missing := ids.Missing(); len(missing) != 0 {
+		return nil, errors.NotFoundf("subnets %v", missing)
+	}
+
+	return results, nil
+}
+
+func (e *environ) getMatchingSubnets(subnetIds IncludeSet, zones []string) ([]network.SubnetInfo, error) {
+	allSubnets, err := e.gce.Subnetworks(e.cloud.Region)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var results []network.SubnetInfo
+	for _, subnet := range allSubnets {
+		if subnetIds.Include(subnet.SelfLink) {
+			results = append(results, makeSubnetInfo(
+				network.Id(subnet.SelfLink),
+				subnet.IpCidrRange,
+				zones,
+			))
+		}
+	}
+	return results, nil
+}
+
+func (e *environ) getInstanceSubnets(inst instance.Id, subnetIds IncludeSet, zones []string) ([]network.SubnetInfo, error) {
+	ifaces, err := e.NetworkInterfaces(inst)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var results []network.SubnetInfo
+	for _, iface := range ifaces {
+		if subnetIds.Include(string(iface.ProviderSubnetId)) {
+			results = append(results, makeSubnetInfo(
+				iface.ProviderSubnetId,
+				iface.CIDR,
+				zones,
+			))
+		}
+	}
+	return results, nil
+}
+
+// NetworkInterfaces implements environs.NetworkingEnviron.
+func (e *environ) NetworkInterfaces(instId instance.Id) ([]network.InterfaceInfo, error) {
+	insts, err := e.Instances([]instance.Id{instId})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	envInst, ok := insts[0].(*environInstance)
+	if !ok {
+		// This shouldn't happen.
+		return nil, errors.Errorf("couldn't extract google instance for %q", instId)
+	}
+	googleInst := envInst.base
+	// According to GCE docs, an instance can only have one nic.
+	// https://cloud.google.com/compute/docs/instances/#instances_and_networks
+	if ifaces := len(googleInst.NetworkInterfaces()); ifaces != 1 {
```

Comment:
My thinking was that it would be a bit of fiddly code that we don't need now - we don't want to do one `Subnets` call for each interface, so it would be something like:
* collect the subnet ids
* make the call and pull the results into a map
* match each interface up with its subnet
* handle and test for the cases where not all the subnets come back.

But I wasn't super comfortable with that decision - like you say, the source is a list so probably there'll be more things in there at some point. I'll expand it out.